[modbus] [modbus_controller] server coil registers

gotnone opened the pull request:
# What does this implement/fix?
Adds Write Single Coil 0x5 command support to the server / slave role of the `modbus_controller`
Adds necessary supporting functionality to the `modbus` component
Adds tests to the `modbus_controller` common test script

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#4825

## Test Environment

- [X] ESP32
- [X] ESP32 IDF

Code gen appears correct, but I don't have hardware to test the following:
- [X] ESP8266
- [X] RP2040
- [X] BK72xx
- [X] RTL87xx

## Example entry for `config.yaml`:

> [!NOTE]
> Using this PR as an external component should pull in both the `modbus` and `modbus_controller` components.
> ```
> external_components:
>   - source: github://pr#8577
>     components: [modbus, modbus_controller]
>```

```yaml
uart:
  - id: uart_modbus_server
    tx_pin: GPIO25
    rx_pin: GPIO35
    baud_rate: 57600

modbus:
  - uart_id: uart_modbus_server
    id: modbus_server
    role: server

modbus_controller:
  - modbus_id: modbus_server
    address: 0x1
    server_coil_registers:
      - address: 0x0002
        write_lambda: |-
          if(state) {
            id(relay1).turn_on();
          } else {
            id(relay1).turn_off();
          }

output:
  - platform: gpio
    id: relay1
    pin: GPIO26
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Commit:
[modbus_controller] Add server role coil registers
[modbus] Add server role coil register support

Commit:
[modbus_controller] Add test server coil register

gotnone reviewed (commented):
This probably needs to be a logical OR with CONF_SERVER_COIL_REGISTERS `modbus_controller/__init__.py:L230`

Commit:
[modbus_controller] fix _final_validate

Commit:
[modbus_controller] fix #6698
`server_register` address ranges may not overlap

gotnone commented:
Probably should integrate PR #8697

Commit:
[modbus_controller] fix error function code

Commit:
Merge remote-tracking branch 'origin/dev' into fix_6698

Commit:
[modbus_controller] prevent duplicate addresses
`server_coil_register` address ranges may not overlap

Commit:
Merge remote-tracking branch 'origin/dev' into server_coil_registers

Commit:
Merge branch 'fix_6698' into server_coil_registers

jmccoy555 commented:
Hi, am I doing something wrong......

```
Compiling .pioenvs/modbus/src/esphome/components/mqtt/mqtt_fan.cpp.o
In file included from src/esphome/components/modbus_controller/modbus_controller.cpp:1:
src/esphome/components/modbus_controller/modbus_controller.h:465:8: error: 'void esphome::modbus_controller::ModbusController::on_modbus_write_coil_register(uint8_t, uint16_t, uint16_t)' marked 'final', but is not virtual
  465 |   void on_modbus_write_coil_register(uint8_t function_code, uint16_t address, uint16_t state) final;
      |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Compiling .pioenvs/modbus/src/esphome/components/mqtt/mqtt_light.cpp.o
Compiling .pioenvs/modbus/src/esphome/components/mqtt/mqtt_lock.cpp.o
*** [.pioenvs/modbus/src/esphome/components/modbus_controller/modbus_controller.cpp.o] Error 1
========================= [FAILED] Took 457.86 seconds =========================
```

Really need this (and ideally the ablility to write holiding registers also).

gotnone commented:
> Hi, am I doing something wrong......
> 
> ```
> Compiling .pioenvs/modbus/src/esphome/components/mqtt/mqtt_fan.cpp.o
> In file included from src/esphome/components/modbus_controller/modbus_controller.cpp:1:
> src/esphome/components/modbus_controller/modbus_controller.h:465:8: error: 'void esphome::modbus_controller::ModbusController::on_modbus_write_coil_register(uint8_t, uint16_t, uint16_t)' marked 'final', but is not virtual
>   465 |   void on_modbus_write_coil_register(uint8_t function_code, uint16_t address, uint16_t state) final;
>       |        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> Compiling .pioenvs/modbus/src/esphome/components/mqtt/mqtt_light.cpp.o
> Compiling .pioenvs/modbus/src/esphome/components/mqtt/mqtt_lock.cpp.o
> *** [.pioenvs/modbus/src/esphome/components/modbus_controller/modbus_controller.cpp.o] Error 1
> ========================= [FAILED] Took 457.86 seconds =========================
> ```
> 
> Really need this (and ideally the ablility to write holiding registers also).

Are you based on dev?

There is more discussion for this and `server_write_registers` on https://discord.com/channels/429907082951524364/1360485495192420492

jmccoy555 commented:
Tried on 2025.5.0-dev and  2025.4.1

gotnone commented:
> Tried on 2025.5.0-dev and 2025.4.1

What does your `external_components:` look like?

Do you include components for `modbus` and `modbus_controller`?  This PR touches both.

I can reproduce your errors if I only grab `modbus_controller` from the PR.

Try something like:

```
external_components:
  - source: github://pr#8577
    components: [modbus, modbus_controller]
```

gotnone commented:
If you need write_registers support I have a branch with that https://github.com/gotnone/esphome@server_write_registers

I haven't made a PR for it yet.  Development is ongoing with discussion at the previous discord link.

jmccoy555 commented:
> > Tried on 2025.5.0-dev and 2025.4.1
> 
> What does your `external_components:` look like?
> 
> Do you include components for `modbus` and `modbus_controller`? This PR touches both.
> 
> I can reproduce your errors if I only grab `modbus_controller` from the PR.
> 
> Try something like:
> 
> ```
> external_components:
>   - source: github://pr#8577
>     components: [modbus, modbus_controller]
> ```

Thanks for the help..... that's done the trick! 

Will have a look at getting the write_registers support working too.

vpiyanov commented:
@gotnone, Thank you for providing this PR!
Do I get it right that it supports only write of coil registers? The read is not supported?

gotnone commented:
@vpiyanov There have been some changes to the modbus code since I created this PR.  I don't know if it will apply cleanly.

Yes the `server_coil_registers` config entry only supports 0x5 `write server coil` for the list of addresses that the user provides.  The user also provides a `write_lambda` which takes a single boolean parameter which is the coil state as parsed from the incoming modbus packet.

If you need the modbus command 0x1 `read_coil` support, that is not currently supported in my PR or in the newest additions to the esphome modbus code as far as I know.

Post back with your needs or open a feature request.  You can also check in on the development channel '#devs' on the esphome discord

Commit:
[modbus] [modbus_controller] server_write_coil

- Merge branch 'dev' into server_coil_registers
- Add set_write_lambda to ServerCoilRegister class

gotnone commented:
@vpiyanov I have just pushed a new version which should work with the newer modbus code in the `dev` branch.  I am not near my test bench, so i have not actually tested this with a serial monitor.

If you try it out, please let me know if the response code sent to the client / master is correct.

The new modbus code also touched the way that modbus errors are reported if the client / master requests a register address that is not configured.  If you get a chance to test that Please let me know if that reports a type 0x1 error correctly as well.

Commit:
[modbus] [modbus_controller] server_write_coil

- Small reversion touching server_registers code

Commit:
[modbus_controller] server_coil_registers

- Check for correct coil state 0x0000 or 0xFF00 in modbus write coil
  command otherwise reply with 0x3 error message

vpiyanov commented:
@gotnone, thank you for updating the code!
Unfortunately I'm far from home now and will be able to try it in the August only.
Will also need to figure out how to deal with reading the coil registers.

Commit:
[modbus_controller] server coil registers

- Satisfy the linter
- Adjust ordering of prototypes in preparation for read coils command

gotnone commented:
@jmccoy555 I am not sure if you saw that a different PR https://github.com/esphome/esphome/pull/9159 added server write registers support was accepted and it is already included in 2025.7.0b beta release.

jmccoy555 commented:
> Hi @gotnone ! I was looking for coil registers read/write support in EspHome and I have noticed the [server_coil_hybrid](https://github.com/gotnone/esphome/tree/server_coil_hybrid) branch in your repository. I have tried it and it worked like a charm! Really cool! Do you plan to create a pull request for it?

Hi @vpiyanov  any chance of sharing  your code as an example please?

gotnone commented:
@jmccoy555 I haven't had a chance to look at this code in a while. Let me see if there is anything on discord to jog my memory.  I recall there were some changes to the server role modbus in the middle of summer.  I thought I updated most of my branches to work at that time.  Problem with being outside of the main development tree is branches get stale.
